lib/ccnet/ccnet.js

```diff
@@ -0,0 +1,208 @@
+  EventEmitter.call(this)
+  this.pollingInterval = null
+  this.device = config.rs232.device
+  this.fiatCode = null
+  this.denominations = null
+  this.billTable = null
+  this._throttledError = _.throttle(2000, err => this.emit('error', err))
+}
+
+util.inherits(Ccnet, EventEmitter)
+Ccnet.factory = function factory (config) {
+  return new Ccnet(config)
+}
+
+Ccnet.prototype.setFiatCode = function (fiatCode) {
+  this.fiatCode = fiatCode
+}
+
+Ccnet.prototype.lightOn = _.noop
+
+Ccnet.prototype.lightOff = _.noop
+
+Ccnet.prototype.run = function (cb) {
+  return rs232.create(this.device)
+    .then(() => {
+      rs232.emitter.on('handleResponse', (response, command) => this._handleResponse(response, command))
+      rs232.emitter.on('error', (err) => this._throttledError(err))
+      rs232.emitter.on('disconnected', () => this.emit('disconnected'))
+      statesFsm.on('command', it => this._send(it))
+      statesFsm.on('billAccepted', () => this.emit('billAccepted'))
+      statesFsm.on('jam', () => this.emit('jam'))
+      statesFsm.on('billValid', () => this.emit('billValid'))
+      statesFsm.on('billRejected', () => this.emit('billRejected'))
+      statesFsm.on('stackerOpen', () => this.emit('stackerOpen'))
+      statesFsm.on('billRead', (data) => {
+        const denomination = this._denominations()[data[0]]
+        if (!denomination) {
+          console.log('bill rejected: unsupported denomination.')
+          this._send(commands.RETURN)
+          return
+        }
+        this.emit('billRead', { denomination })
+      })
+    })
+    .then(() => {
+      this._startPolling()
+      this._connect()
+
+      const t0 = Date.now()
+      const denominationsInterval = setInterval(() => {
+        if (this.hasDenominations()) {
+          clearInterval(denominationsInterval)
+          return cb()
+        }
+
+        if (Date.now() - t0 > 5000) {
+          clearInterval(denominationsInterval)
+          cb(new Error('Timeout waiting for denominations'))
+        }
+      }, 500)
+    })
+    .catch(cb)
+}
+
+Ccnet.prototype.close = function (cb) {
+  this._stopPolling()
+  rs232.close(function (err) {
+    if (err) console.log(err)
+    cb(err)
+  })
+}
+
+Ccnet.prototype.getBillsToEnable = function () {
+  // at this point this enables every bill available on the firmware
+  // take fiatCode into account once we do that in id003 as well
+  const billsBuffer = Buffer.alloc(3)
+  const billsToEnable = _.keys(this._denominations())
+
+  billsToEnable.forEach(it => {
+    writeBit(billsBuffer, 2 - Math.floor(it / 8), it % 8, 1)
+  })
+
+  return Buffer.concat([billsBuffer, billsBuffer])
+}
+
+Ccnet.prototype.enable = function () {
+  const billsBuffer = this.getBillsToEnable()
+
+  const command = Buffer.concat([Buffer.from(commands.ENABLE_BILL_TYPES), billsBuffer])
+  this._send(command)
+}
+
+Ccnet.prototype.disable = function () {
+  const command = Buffer.concat([Buffer.from(commands.ENABLE_BILL_TYPES), Buffer.alloc(6)])
```

Comment:
Maybe *commands* should already be buffers, then we won't have to do `Buffer.from` each time.

Reply:
Is this for performance reasons or styling? I'm currently using commands without Buffer.from more often then not on code, but at the end rs232 code always redoes it.

Reply:
Just to make the code more compact. If all the commands are predefined as buffers it should make the code more simple.